File: python/patterns-use-cases/src/queue/async_task_worker.py
import restate
from pydantic import BaseModel
from restate import Service, Context


class TaskOpts(BaseModel):
    id: str


async_task_worker = Service("AsyncTaskWorker")


@async_task_worker.handler()
async def run(ctx: Context, params: TaskOpts):
    # ... some heavy work ...
    return f"Finished work on task: {params.id}"


app = restate.app([async_task_worker])

File: typescript/patterns-use-cases/src/queue/async_task_worker.ts
import * as restate from "@restatedev/restate-sdk";
import { Context } from "@restatedev/restate-sdk";

const asyncTaskWorker = restate.service({
  name: "asyncTaskWorker",
  handlers: {
    runTask: async (ctx: Context, params: TaskOpts) => {
      return someHeavyWork(params);
    },
  },
});

export type AsyncTaskWorker = typeof asyncTaskWorker;

restate.endpoint().bind(asyncTaskWorker).listen(9080);


// ----------------------- Stubs to please the compiler -----------------------
export type TaskOpts = {id: string, task: string};

function someHeavyWork(work: TaskOpts) {
  return "Work!";
}
